http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/SocketOptionsConfiguration.java

```diff
@@ -0,0 +1,215 @@
+    }
+
+    /**
+     * @return socket domain
+     */
+    public SocketOptions.SocketDomain domain() {
+        return domain;
+    }
+
+    /**
+     * @return socket type
+     */
+    public SocketOptions.SocketType type() {
+        return type;
+    }
+
+    /**
+     * @return number of milliseconds before a connection will be considered timed out
+     */
+    public int connectTimeoutMs() {
+        return connectTimeoutMs;
+    }
+
+    /**
+     * @return number of seconds between TCP keepalive packets being sent to the peer
+     */
+    public int keepAliveIntervalSecs() {
+        return keepAliveIntervalSecs;
+    }
+
+    /**
+     * @return number of seconds to wait for a keepalive response before considering the connection timed out
+     */
+    public int keepAliveTimeoutSecs() {
+        return keepAliveTimeoutSecs;
+    }
+
+    public static Builder builder() {
+        return new DefaultSocketOptionsConfigurationBuilder();
+    }
+
+    /**
+     * A builder for {@link SocketOptionsConfiguration}.
+     *
+     * <p>All implementations of this interface are mutable and not thread safe.</p>
+     */
+    public interface Builder {
+
+        /**
+         * Sets the socket domain
+         * @param domain socket domain
+         * @return Builder
+         */
+        Builder domain(SocketOptions.SocketDomain domain);
+
+        /**
+         * Sets the socket type
+         * @param type socket type
+         * @return Builder
+         */
+        Builder type(SocketOptions.SocketType type);
+
+        /**
+         * Sets the number of milliseconds before a connection will be considered timed out
+         * @param connectTimeoutMs number of milliseconds before a connection will be considered timed out
+         * @return Builder
+         */
+        Builder connectTimeoutMs(int connectTimeoutMs);
+
+        /**
+         * Sets the number of seconds between TCP keepalive packets being sent to the peer
+         * @param keepAliveIntervalSecs number of seconds between TCP keepalive packets being sent to the peer
+         * @return Builder
+         */
+        Builder keepAliveIntervalSecs(int keepAliveIntervalSecs);
+
+        /**
+         * Sets the number of seconds to wait for a keepalive response before considering the connection timed out
+         * @param keepAliveTimeoutSecs number of seconds to wait for a keepalive response before considering the connection timed out
+         * @return Builder
+         */
+        Builder keepAliveTimeoutSecs(int keepAliveTimeoutSecs);
+
+        SocketOptionsConfiguration build();
+    }
+
+    /**
+     * An SDK-internal implementation of {@link Builder}.
+     */
+    private static final class DefaultSocketOptionsConfigurationBuilder implements Builder {
+
+        private SocketOptions.SocketDomain domain = SocketOptions.SocketDomain.IPv6;
+        private SocketOptions.SocketType type = SocketOptions.SocketType.STREAM;
+        private int connectTimeoutMs = 3000;
```

Comment:
I discussed this with @millems and we thought that using the 2 sec default of the Global Http Default would be a better user experience than the 3 sec hard coded value hard coded in the `SocketOption` class. So I would use the value in `SdkHttpConfigurationOption.GLOBAL_HTTP_DEFAULTS`.

Reply:
CC @zoewangg `SdkHttpConfigurationOption.DEFAULT_CONNECTION_TIMEOUT` is 2 seconds, but an unconfigured SocketOptions currently defaults to 3 seconds.

Reply:
OK so I guess the real question is whether `SocketOptionsConfiguration` should have a field for `connectionTimeout` at all or if it should come from `SdkHttpConfigurationOption.CONNECTION_TIMEOUT`.

Or is that where all the KeepAlive settings should move to as well instead of a dedicated object?

Reply:
Yeah, they should come from the SDK predefined defaults. We can just add the configured value to the standard options map in the builder setter methods. See
https://github.com/aws/aws-sdk-java-v2/pull/3457#discussion_r987157558